### PR TITLE
BUG: Join with a list of a single element should behave as a join with a single element (#57676)

### LINKED ISSUE #57676: BUG: `join` with `list` does not behave like singleton
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np

cat_df = pd.DataFrame({'cat': pd.Categorical(['a', 'v', 'd'])}, index=pd.Index(['a', 'b', 'c'], name='y'))
join_df = pd.DataFrame({'foo': np.arange(6)}, index = pd.MultiIndex.from_tuples([(0, 'a'), (0, 'b'), (0, 'c'), (1, 'a'), (1, 'b'), (1, 'c')], names=('x', 'y')) )
join_df.join([cat_df]) # NaNs in the `cat` column
join_df.join(cat_df) # correct
```

### Issue Description

I would expect the result to be identical.  I am really interested in being able to left join multiple `cat_df`s that share (one of) an index with the `join_df`.

### Expected Behavior

I would expect identical behavior.  I did notice that with the `on` `kwarg`, I get an error that might indicate this is not allowed:

```python
join_df.join([cat_df], on='y')
```
gives
```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/ilangold/Projects/Theis/pandas/pandas/core/frame.py", line 10463, in join
    raise ValueError(
ValueError: Joining multiple DataFrames only supported for joining on index
```

### Installed Versions

Not sure what's up with the installed version commit since my `git log` looks like

```
commit e14a9bd41d8cd8ac52c5c958b735623fe0eae064 (HEAD -> main, origin/main, origin/HEAD)
Author: Eric Larson <larson.eric.d@gmail.com>
Date:   Wed Feb 28 17:21:32 2024 -0500

    ENH: Report how far off the formatting is (#57667)

```
<details>

INSTALLED VERSIONS
------------------
commit                : 52cb549f443f09727448251cfee53227c6bca9d2
python                : 3.11.6.final.0
python-bits           : 64
OS                    : Darwin
OS-release            : 22.6.0
Version               : Darwin Kernel Version 22.6.0: Wed Oct  4 21:26:23 PDT 2023; root:xnu-8796.141.3.701.17~4/RELEASE_ARM64_T6000
machine               : arm64
processor             : arm
byteorder             : little
LC_ALL                : None
LANG                  : None
LOCALE                : None.UTF-8

pandas                : 3.0.0.dev0+87.g52cb549f44.dirty
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.8.2
setuptools            : 68.2.2
pip                   : 23.3.1
Cython                : 3.0.8
pytest                : 8.0.2
hypothesis            : 6.98.13
sphinx                : 7.2.6
blosc                 : None
feather               : None
xlsxwriter            : 3.2.0
lxml.etree            : 5.1.0
html5lib              : 1.1
pymysql               : 1.4.6
psycopg2              : 2.9.9
jinja2                : 3.1.3
IPython               : 8.22.1
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : 4.12.3
bottleneck            : 1.3.8
fastparquet           : 2024.2.0
fsspec                : 2024.2.0
gcsfs                 : 2024.2.0
matplotlib            : 3.8.3
numba                 : 0.59.0
numexpr               : 2.9.0
odfpy                 : None
openpyxl              : 3.1.2
pyarrow               : 15.0.0
pyreadstat            : 1.2.6
python-calamine       : None
pyxlsb                : 1.0.10
s3fs                  : 2024.2.0
scipy                 : 1.12.0
sqlalchemy            : 2.0.27
tables                : 3.9.2
tabulate              : 0.9.0
xarray                : 2024.2.0
xlrd                  : 2.0.1
zstandard             : 0.22.0
tzdata                : 2024.1
qtpy                  : None
pyqt5                 : None

</details>

### PR DESCRIPTION
**Pull request by Dacops**
- [x] closes #57676
- [x] Added a test to pandas/tests/frame/methods/test_join.py 

Even though it does not make much sense, joining a dataframe with a list of a single element should behave as joining the dataframe with that element.

### CODE REVIEW

**Comment by Dacops** [comment]
Seems like the same test is failing over and over again, however it does not make sense to me, the test is:
`def test_suffix_on_list_join():
        first = DataFrame({"key": [1, 2, 3, 4, 5]})
        second = DataFrame({"key": [1, 8, 3, 2, 5], "v1": [1, 2, 3, 4, 5]})
        third = DataFrame({"keys": [5, 2, 3, 4, 1], "v2": [1, 2, 3, 4, 5]})
    
        # check proper errors are raised
        msg = "Suffixes not supported when joining multiple DataFrames"
        with pytest.raises(ValueError, match=msg):
>           first.join([second], lsuffix="y")
E           Failed: DID NOT RAISE <class 'ValueError'>`

In this example multiple DataFrames are not passed, just a list with a single DataFrame

**Comment by Dacops** [comment]
I've updated the test mentioned above, before any list passed with the Dataframe.join method was assumed to have 2+ elements, by adding this edge case where a list can have a single element (behaving has a Dataframe.join with a single element) that condition in the test "test_suffix_on_list_join" no longer makes sense.

File: pandas/tests/frame/methods/test_join.py [hunk 1]
```diff
@@ -562,3 +560,21 @@ def test_frame_join_tzaware(self):
...
+            index=MultiIndex.from_tuples(
+                [(0, "a"), (0, "b"), (0, "c"), (1, "a"), (1, "b"), (1, "c")],
+                names=("x", "y"),
+            ),
+        )
+
+        result = test2.join([test1])
+        expected = test2.join(test1)
+
+        assert result.equals(expected)
```
Comment (Aloqeely) [inline pandas/tests/frame/methods/test_join.py:580]: You should use `tm.assert_frame_equal` to 2 DataFrames

**Comment by Aloqeely** [comment]
LGTM. @mroeschke would you mind taking a look?

**Comment by Dacops** [comment]
Thanks for the review @Aloqeely, in the meantime I've rebased my PR to address conflicts in the "whatsnew" file. I've noticed that the PR started failing the Numpy Dev tests because it was rebased from the main Pandas branch which is failing the same tests. No other changes (except the whatsnew file) were made since the last review. If any developer familiar with this area of the code could take a look, I'd appreciate it (cc @mroeschke, @WillAyd, @jbrockmendel ) I apologize for the broader ping, but I wasn't sure who would be best suited to review this change.

**Comment by WillAyd** [comment]
Hmm unfortunately I am -1 on making this change. I'm not sure I understand why we need to special case a single element list.

Generally our requirement for joins / merges is that the key be hashable

**Comment by WillAyd** [comment]
Sorry ignore my hashability comment - I see this is in regards to the join argument not the key elements themselves. Still, I'm not sure why we need to special case this

**Comment by Dacops** [comment]
There's a problem, in some rare cases, that joining a dataframe x with a list containing a single dataframe y would lose information in relation to just joining dataframe x with dataframe y (one example of this on the original issue: https://github.com/pandas-dev/pandas/issues/57676)

**Comment by WillAyd** [comment]
Ah OK thanks @Dacops . This only solves that then though with a single element list right? What happens when there are multiple list elements? I think there is a fix somewhere else in the code that would work generally - ideally we avoid special-casing solutions like this as they don't add up well over time

**Comment by Dacops** [comment]
With multiple list elements the behaviour remains the same. Basically, what the code did was if the joining element is an Object deal with it in the way A, if it's a List deal with it in the way B. I've added that if-statemen that just checks if the list has a single element. If it does use way A else keep using way B.

**Comment by Dacops** [comment]
Hmmm, yeah I did thought of that, however the line that handles that is on frame.py/10685: `can_concat = all(df.index.is_unique for df in frames)` and it is been there for 12 years. Therefore I assumed the logic was correct since it wasn't changed for that long and it only fails in very rare cases. I thought it would be more prudent to add the if-statement for the edge cases instead of changing code that has been unaltered for so long.

**Comment by Aloqeely** [comment]
It might very well be broken logic even if it was there for 12 years, funnily enough I just saw [another comment](https://github.com/pandas-dev/pandas/pull/57494#discussion_r1587386787) saying that some broken logic was introduced 13 years ago

**Comment by Dacops** [comment]
I think I've identified the problem, that [expression](https://github.com/pandas-dev/pandas/blob/ac65c1391b3a07423c21fcb96c380bb2ae51b677/pandas/core/frame.py#L10685) evaluates if all inserted indexes from all objects in the list are unique so that they can be concatenated (stacked on top of each other). However if a MultiIndex is used the entire MultiIndex is compared not separated indexes inside it. A problem can arise here, for example on the original [issue](https://github.com/pandas-dev/pandas/issues/57676) a Indexed dataframe is getting joined to a MultiIndexed dataframe, the pairs of indexes of the MultiIndexed dataframe are compared to the indexes of the Indexed dataframe. There's no repetition, so they get concatenated, but here's the error. One of the MultiIndexed dataframes index is 'y' same as the index of the Indexed dataframe, and those no longer are unique, which means they need to be merged. Solution I'm thinking of, after the lambda function to check if all indexes are unique, if this evaluates to True add an extra check to look for repeated indexes and if any of these are not unique change the value to False and skip to the Merge instead of a Concat.

**Comment by WillAyd** [comment]
Hmm I'm not sure I follow - I feel like the indexes should be the same when trying to join. Assuming some kind of matching behavior between an Index and MultiIndex might have a lot of logic pitfalls

**Comment by Dacops** [comment]
The problem is with the MultiIndexes, for example in the original [issue](https://github.com/pandas-dev/pandas/issues/57676), the Index is `index=pd.Index(['a', 'b', 'c'], name='y'))`, while the MultiIndex is `index=pd.MultiIndex.from_tuples([(0, 'a'), (0, 'b'), (0, 'c'), (1, 'a'), (1, 'b'), (1, 'c')], names=('x', 'y'))`. When `all(df.index.is_unique for df in frames)` is run over both it gives out the values: `'a', 'b', 'c', (0, 'a'), (0, 'b'), (0, 'c'), (1, 'a'), (1, 'b'), (1, 'c')` which returns True (there are no repetitions, value of indexes are unique, concatenate them). However this cannot happen since the MultiIndex has a 'y' index, if we extract it and repeat `all(df.index.is_unique for df in frames)` over both 'y' we get `'a', 'b', 'c', 'a', 'b', 'c', 'a', 'b', 'c'` which returns False (there are repeated values, we can't concatenate them). Tl;dr when a MultiIndex is passed it's getting evaluated as a whole and not the individual indexs inside it where the merge may occur.

**Comment by Dacops** [comment]
I believe I've got the solution for this, I'll push it so you can take a look at it

**Comment by Aloqeely** [comment]
So is this a problem with `is_unique` on `MultiIndex`? Do you think it's current behavior makes sense? I think it should return True only if each level has unique values, and if I understood you correctly that seems to be failing right now

If you agree with my suggested behavior then I will open an issue to fix it, if not we should at least document how it operates on `MultiIndex` (don't see any examples for that)

**Comment by Dacops** [comment]
@Aloqeely I believe the current behaviour of `is_unique` for MultiIndex is correct, however for this case this behaviour would not work. This because we're joining an Indexed dataframe with a MultiIndexed dataframe, the join will only occur in a single index (the one of the Indexed dataframe), so the remaining indexs of the MultiIndex are irrelevant to this operation.

**Comment by Dacops** [comment]
this doc build and upload check is something I did wrong or do I just re-run the checks, I haven't touched that `/home/runner/work/pandas/pandas/doc/source/user_guide/merging.rst`

**Comment by Aloqeely** [comment]
It does seem related to your changes, I've summarized the example in the doc that fails:
```python
import pandas as pd

left = pd.DataFrame(
       {
           "A": ["A0", "A1", "A2", "A3"],
           "B": ["B0", "B1", "B2", "B3"],
           "key2": ["K0", "K1", "K0", "K1"],
       },
       index=pd.Index(["K0", "K0", "K1", "K2"], name="key1"),
   )

right = pd.DataFrame({"v": [7, 8, 9]}, index=["K1", "K1", "K2"])

result = left.join([right])
```
Raises
```
UnboundLocalError: cannot access local variable 'can_concat' where it is not associated with a value
```
It is surprising that only the doc build failed but not any tests, so once you fix it for this case please add tests.

**Comment by WillAyd** [comment]
We are fixing the wrong thing from the original issue. I think specifying `on=...` is reasonable to allow an Index and a MultiIndex to be joined together, but we should not be doing anything to automatically try and determine this. We should actually be raising when the indexes have a different number of levels in join

**Comment by Dacops** [comment]
Oh thanks @Aloqeely , I figured out what it was,  on this [line](https://github.com/pandas-dev/pandas/blob/04ac369326a4a3211e60cfc31df85f9c61a19933/pandas/core/frame.py#L10687), in case "common_indexes" was empty can_concat would not be initialized and it is called in the following if statement, so it threw an error there. I initialized it to a default value before the for loop so that doesn't occur anymore. Everything's good now, @WillAyd if you could re-review the PR I would appreciate it, thanks.

**Comment by WillAyd** [comment]
Right if you are trying to join two index objects with non-equal levels

**Comment by Dacops** [comment]
Forgot the default value need to be there to avoid the problem with the doc build mentioned above, both the raise and default value are there now

**Comment by Dacops** [comment]
wait it isn't that, @Aloqeely how can I run the doc build and upload test locally to see what's failing

**Comment by Dacops** [comment]
@Aloqeely how can I get better errors on the Doc build and upload test like you sent instead of just `RuntimeError: Unexpected exception in `/home/runner/work/pandas/pandas/doc/source/user_guide/merging.rst` line 930` ?

**Comment by Aloqeely** [comment]
You can open the doc file which has the errors and go to the error line, run the examples near to that line manually until you see which example causes the error.

**Comment by Dacops** [comment]
Oh thanks @Aloqeely, so if my error says `RuntimeError: Unexpected exception in /home/runner/work/pandas/pandas/doc/source/user_guide/merging.rst line 930`, [here](https://github.com/pandas-dev/pandas/blob/e0d30615a0450f95ba89c06a4f0fd6f017259085/doc/source/user_guide/merging.rst?plain=1#L930). Then the test that fails is the one above? The one that starts on line 906, [here](https://github.com/pandas-dev/pandas/blob/e0d30615a0450f95ba89c06a4f0fd6f017259085/doc/source/user_guide/merging.rst?plain=1#L906)?

**Comment by Aloqeely** [comment]
I think so

**Comment by Dacops** [comment]
I know what's happening, in this [example](https://pandas.pydata.org/docs/user_guide/merging.html#joining-multiple-dataframe) left and right both have an index named 'key1' while right2 has an index but it is not named. This not named index is inferred to be 'key1', is this expected behaviour? Taking in consideration your input @WillAyd ?

> We are fixing the wrong thing from the original issue. I think specifying `on=...` is reasonable to allow an Index and a MultiIndex to be joined together, but we should not be doing anything to automatically try and determine this. We should actually be raising when the indexes have a different number of levels in join

**Comment by Dacops** [comment]
I've changed the test in the documentation that was failing, instead of joining index "key 1" on "key 1" on None now they're all "key 1"

**Comment by Dacops** [comment]
@WillAyd the issue should be completed now

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**